File: backend/src/services/ConfigService.ts
import { prisma } from '../db/client.js';
import { CommitService } from './CommitService.js';

export interface CommitDateRange {
  earliest: number;  // 最早提交时间（毫秒）
  latest: number;    // 最晚提交时间（毫秒）
}

export interface RepositoryConfig {
  id: string;
  name: string;
  path: string;
  enabled: boolean;
  authors: AuthorConfig[];
  isAbnormal: boolean;
  abnormalReason?: string;
  lastScanTime: number | null;
  totalCommits: number;
  commitDateRange: CommitDateRange | null;  // 提交时间范围，无记录时为 null
  createdAt: number;
  updatedAt: number;
}

export interface AuthorConfig {
  id: number;
  name: string;
  email: string;
  isDefault: boolean;
}

const ABNORMAL_REPOSITORIES_SETTING_KEY = 'abnormal_repositories';

export class ConfigService {
  private commitService: CommitService;

  constructor() {
    this.commitService = new CommitService();
  }

  private async getAbnormalRepositoryMarks(): Promise<Record<string, string>> {
    const setting = await prisma.appSetting.findUnique({
      where: { key: ABNORMAL_REPOSITORIES_SETTING_KEY }
    });

    if (!setting?.value) return {};

    try {
      const parsed = JSON.parse(setting.value);
      if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
        return {};
      }
      return parsed as Record<string, string>;
    } catch {
      return {};
    }
  }

  async setAbnormalRepositoryMarks(items: Array<{ id: string; reason: string }>): Promise<void> {
    const now = BigInt(Date.now());
    const marks = items.reduce<Record<string, string>>((acc, item) => {
      const id = item.id.trim();
      const reason = item.reason.trim();
      if (id && reason) {
        acc[id] = reason;
      }
      return acc;
    }, {});

    await prisma.appSetting.upsert({
      where: { key: ABNORMAL_REPOSITORIES_SETTING_KEY },
      create: {
        key: ABNORMAL_REPOSITORIES_SETTING_KEY,
        value: JSON.stringify(marks),
        updatedAt: now
      },
      update: {
        value: JSON.stringify(marks),
        updatedAt: now
      }
    });
  }

  /**
   * 获取仓库完整配置（含作者）
   */
  async getRepositoryConfig(repoId: string): Promise<RepositoryConfig | null> {
    const repo = await prisma.repository.findUnique({
      where: { id: repoId },
      include: {
        authors: {
          orderBy: { createdAt: 'asc' }
        }
      }
    });

    if (!repo) return null;

    const commitDateRange = await this.commitService.getCommitDateRangeForRepo(repoId);
    const abnormalMarks = await this.getAbnormalRepositoryMarks();
    const abnormalReason = abnormalMarks[repo.id];

    return {
      id: repo.id,
      name: repo.name,
      path: repo.path,
      enabled: repo.enabled,
      isAbnormal: Boolean(abnormalReason),
      abnormalReason,
      authors: repo.authors.map(a => ({
        id: a.id,
        name: a.name,
        email: a.email,
        isDefault: a.isDefault
      })),
      lastScanTime: repo.lastScanTime ? Number(repo.lastScanTime) : null,
      totalCommits: repo.totalCommits,
      commitDateRange,
      createdAt: Number(repo.createdAt),
      updatedAt: Number(repo.updatedAt)
    };
  }

  /**
   * 获取所有仓库配置
   */
  async getAllRepositoriesConfig(): Promise<RepositoryConfig[]> {
    const repos = await prisma.repository.findMany({
      include: {
        authors: {
          orderBy: { createdAt: 'asc' }
        }
      },
      orderBy: { name: 'asc' }
    });

    // 并行获取所有仓库的提交时间范围
    const commitDateRanges = await Promise.all(
      repos.map(repo => this.commitService.getCommitDateRangeForRepo(repo.id))
    );
    const abnormalMarks = await this.getAbnormalRepositoryMarks();

    return repos.map((repo, index) => {
      const abnormalReason = abnormalMarks[repo.id];

      return {
        id: repo.id,
        name: repo.name,
        path: repo.path,
        enabled: repo.enabled,
        isAbnormal: Boolean(abnormalReason),
        abnormalReason,
        authors: repo.authors.map(a => ({
          id: a.id,
          name: a.name,
          email: a.email,
          isDefault: a.isDefault
        })),
        lastScanTime: repo.lastScanTime ? Number(repo.lastScanTime) : null,
        totalCommits: repo.totalCommits,
        commitDateRange: commitDateRanges[index],
        createdAt: Number(repo.createdAt),
        updatedAt: Number(repo.updatedAt)
      };
    });
  }

  /**
   * 获取仓库的作者列表
   */
  async getAuthorsByRepoId(repoId: string): Promise<AuthorConfig[]> {
    const authors = await prisma.author.findMany({
      where: { repoId },
      orderBy: { createdAt: 'asc' }
    });

    return authors.map(a => ({
      id: a.id,
      name: a.name,
      email: a.email,
      isDefault: a.isDefault
    }));
  }

  /**
   * 获取所有启用的仓库
   */
  async getEnabledRepositories(): Promise<Array<{ id: string; name: string; path: string; enabled: boolean }>> {
    const repos = await prisma.repository.findMany({
      where: { enabled: true },
      select: {
        id: true,
        name: true,
        path: true,
        enabled: true
      },
      orderBy: { name: 'asc' }
    });

    return repos;
  }

  /**
   * 获取所有仓库的作者邮箱（用于扫描时过滤）
   * 如果仓库没有配置作者，返回空数组
   */
  async getAllAuthorEmails(): Promise<string[]> {
    const authors = await prisma.author.findMany({
      select: { email: true }
    });

    // 去重
    return Array.from(new Set(authors.map(a => a.email)));
  }

  /**
   * 获取指定仓库的作者邮箱列表
   */
  async getAuthorEmailsByRepoId(repoId: string): Promise<string[]> {
    const authors = await prisma.author.findMany({
      where: { repoId },
      select: { email: true }
    });

    return authors.map(a => a.email);
  }

}

File: golang-backend/internal/service/config_service.go
package service

import (
	"github.com/huangjin/coding-history/golang-backend/internal/dto"
	"github.com/huangjin/coding-history/golang-backend/internal/model"
	"gorm.io/gorm"
)

type ConfigService struct {
	db            *gorm.DB
	commitSvc     *CommitService
}

func NewConfigService(db *gorm.DB) *ConfigService {
	return &ConfigService{
		db:        db,
		commitSvc: NewCommitService(db),
	}
}

func (s *ConfigService) GetRepositoryConfig(repoID string) (*dto.RepositoryConfig, error) {
	var repo model.Repository
	err := s.db.Preload("Authors", func(db *gorm.DB) *gorm.DB {
		return db.Order("created_at ASC")
	}).Where("id = ?", repoID).First(&repo).Error
	if err != nil {
		return nil, err
	}

	dateRange, _ := s.commitSvc.GetCommitDateRangeForRepo(repoID)

	return &dto.RepositoryConfig{
		ID:              repo.ID,
		Name:            repo.Name,
		Path:            repo.Path,
		Enabled:         repo.Enabled,
		LastScanTime:    repo.LastScanTime,
		TotalCommits:    repo.TotalCommits,
		CommitDateRange: dateRange,
		CreatedAt:       repo.CreatedAt,
		UpdatedAt:       repo.UpdatedAt,
		Authors:         mapAuthors(repo.Authors),
	}, nil
}

func (s *ConfigService) GetAllRepositoriesConfig() ([]dto.RepositoryConfig, error) {
	var repos []model.Repository
	err := s.db.Preload("Authors", func(db *gorm.DB) *gorm.DB {
		return db.Order("created_at ASC")
	}).Order("name ASC").Find(&repos).Error
	if err != nil {
		return nil, err
	}

	configs := make([]dto.RepositoryConfig, len(repos))
	for i, repo := range repos {
		dateRange, _ := s.commitSvc.GetCommitDateRangeForRepo(repo.ID)
		configs[i] = dto.RepositoryConfig{
			ID:              repo.ID,
			Name:            repo.Name,
			Path:            repo.Path,
			Enabled:         repo.Enabled,
			LastScanTime:    repo.LastScanTime,
			TotalCommits:    repo.TotalCommits,
			CommitDateRange: dateRange,
			CreatedAt:       repo.CreatedAt,
			UpdatedAt:       repo.UpdatedAt,
			Authors:         mapAuthors(repo.Authors),
		}
	}
	return configs, nil
}

func (s *ConfigService) GetAuthorsByRepoID(repoID string) ([]dto.AuthorConfig, error) {
	var authors []model.Author
	err := s.db.Where("repo_id = ?", repoID).Order("created_at ASC").Find(&authors).Error
	if err != nil {
		return nil, err
	}
	return mapAuthorsToDTO(authors), nil
}

func (s *ConfigService) GetEnabledRepositories() ([]dto.EnabledRepository, error) {
	var repos []model.Repository
	err := s.db.Select("id, name, path, enabled").Where("enabled = ?", true).Order("name ASC").Find(&repos).Error
	if err != nil {
		return nil, err
	}
	result := make([]dto.EnabledRepository, len(repos))
	for i, r := range repos {
		result[i] = dto.EnabledRepository{
			ID:      r.ID,
			Name:    r.Name,
			Path:    r.Path,
			Enabled: r.Enabled,
		}
	}
	return result, nil
}

func (s *ConfigService) GetAuthorEmailsByRepoID(repoID string) ([]string, error) {
	var authors []model.Author
	err := s.db.Select("email").Where("repo_id = ?", repoID).Find(&authors).Error
	if err != nil {
		return nil, err
	}
	emails := make([]string, len(authors))
	for i, a := range authors {
		emails[i] = a.Email
	}
	return emails, nil
}

func (s *ConfigService) GetAllAuthorEmails() ([]string, error) {
	var authors []model.Author
	err := s.db.Select("email").Find(&authors).Error
	if err != nil {
		return nil, err
	}
	seen := make(map[string]bool)
	var result []string
	for _, a := range authors {
		if !seen[a.Email] {
			seen[a.Email] = true
			result = append(result, a.Email)
		}
	}
	return result, nil
}

func mapAuthors(authors []model.Author) []dto.AuthorConfig {
	result := make([]dto.AuthorConfig, len(authors))
	for i, a := range authors {
		result[i] = dto.AuthorConfig{
			ID:        a.ID,
			Name:      a.Name,
			Email:     a.Email,
			IsDefault: a.IsDefault,
		}
	}
	return result
}

func mapAuthorsToDTO(authors []model.Author) []dto.AuthorConfig {
	return mapAuthors(authors)
}
